middleman_protocol/middleman_protocol/message.py

```diff
@@ -0,0 +1,196 @@
+
+    @classmethod
+    def _validate_length(cls, message):
+        if not len(message.signed_part_of_the_frame.payload) == message.signed_part_of_the_frame.payload_length:
+            raise MiddlemanProtocolError(
+                f'Deserialized message payload length is {len(message.signed_part_of_the_frame.payload)} '
+                f'instead of {message.signed_part_of_the_frame.payload_length}.'
+            )
+
+    @classmethod
+    def _validate_signature(cls, message):
+        if not message.frame_signature:  # TODO: Really verify signature
+            raise MiddlemanProtocolError(
+                f'Deserialized message signature wrong.'
+            )
+
+    def serialize(self) -> bytes:
+        assert isinstance(self.payload, bytes)
+
+        message_format = self.get_message_format()
+
+        signed_part_of_the_frame = Container(
+            request_id=self.request_id,
+            payload_length=len(self.payload),
+            payload_type=self.payload_type,
+            payload=self.payload,
+        )
+
+        raw_signed_part_of_the_frame = message_format.signed_part_of_the_frame.build(
+            signed_part_of_the_frame
+        )
+
+        #frame_signature = sha256(raw_signed_part_of_the_frame).hexdigest().encode()  # TODO: TMP
+
+        frame_signature = b'x' * 64
+
+        message = Container(
+            frame_signature=frame_signature,
+            signed_part_of_the_frame=signed_part_of_the_frame,
+        )
+
+        raw_message = message_format.build(message)
+        return raw_message
+
+    def _serialize_payload(self, payload):
+        return payload
+
+    def _validate_request_id(self, request_id):
+        if not isinstance(request_id, int):
+            raise MiddlemanProtocolError(
+                f'request_id  is {type(request_id)} instead of int.'
+            )
+
+    def _validate_payload(self, payload):
+        pass
+
+
+class GolemMessageMiddlemanMessage(AbstractMiddlemanMessage):
+
+    payload_type = PayloadType.GOLEM_MESSAGE
+
+    @classmethod
+    def _deserialize_payload(cls, payload: bytes) -> BaseGolemMessage:
+        return BaseGolemMessage.deserialize(payload, None, check_time=False)
+
+    def _serialize_payload(self, payload: BaseGolemMessage) -> bytes:
+        return payload.serialize()
+
+    def _validate_payload(self, payload: BaseGolemMessage) -> None:
+        if not isinstance(payload, BaseGolemMessage):
+            raise MiddlemanProtocolError(
+                f'Trying to create GolemMessageMiddlemanMessage but passed payload type is {type(payload)} '
+                f'instead of Golem Message.'
+            )
+
+
+class ErrorMiddlemanMessage(AbstractMiddlemanMessage):
+
+    payload_type = PayloadType.ERROR
+
+    @classmethod
+    def _deserialize_payload(cls, payload: bytes) -> tuple:
+        return tuple(payload.decode().split('---'))
+
+    def _serialize_payload(self, payload: tuple) -> bytes:
+        return ('---'.join(payload)).encode()
```

Comment:
If you do this then you should also ensure that the error message never contains `---` or you should escape it.

Reply:
It was done this way only for prove of concept purposes. The question is - how `ERROR` payload type originally passed will look like ? I have assumed it will be a tuple, is that correct ?

Reply:
Payload in that case is a pair that contains the error code and the error message. The details are up to you. You have to make a decision about:
1. The content: whether the code is a number or a string and whether the code or the string comes first.
2. The serialization format: whether you use a header with lengths, assume constant length or use a separator.

Those are low level details that can be chosen during implementation. There are downsides and upsides to various solutions but whatever you choose, as long as we can send a code and an error message, it's going to be consistent with the specification.

If you're asking me:
1. I'd go with an int for the code and send code first. I'd also specifically require a pair for this message type, not a tuple of arbitrary length.
2. I'd probably go for a 2- or 4-byte code field followed by a string that takes up the rest of the payload.

Using a separator is also fine, just make sure that you either escape the string and the code or disallow using the separator inside the code (if it's a string) and split only on the first occurrence.

Reply:
:+1: done